libraries/lts/include/mcrl2/lts/detail/liblts_bisim_m.h

```diff
@@ -0,0 +1,579 @@
+#include "mcrl2/lts/lts_utilities.h"
+#include "mcrl2/lts/detail/liblts_scc.h"
+#include "mcrl2/lts/detail/liblts_merge.h"
+#include "mcrl2/lts/lts_aut.h"
+#include "mcrl2/lts/lts_fsm.h"
+#include "mcrl2/lts/lts_dot.h"
+
+namespace mcrl2
+{
+namespace lts
+{
+namespace detail
+{
+template < class LTS_TYPE>
+class bisim_partitioner_counter_example {
+public:
+    /** \brief Creates a bisimulation partitioner for an LTS.
+    *  \details TODO.
+    *  \warning Might be less efficient than other implementations, focussed on generating minimal depth counter-examples.
+    *  \param[in] l Reference to the LTS. */
+    bisim_partitioner_counter_example(LTS_TYPE& l, const std::size_t init_l2)
+        : aut(l),
+        initial_l2(init_l2),
+        max_state_index(0)
+    {
+        create_initial_partition();
+        bool splitted = true;
+        level_type lvl = 1;
+        while (splitted && block_index_of_a_state[initial_l2] == block_index_of_a_state[aut.initial_state()])
+        {
+            splitted = refine_partition(lvl);
+            lvl++; 
+            for (typename std::vector<block>::reverse_iterator i = blocks.rbegin();
+                i != blocks.rend() && (*i).level == lvl - 1; ++i) {
+                to_be_processed.push_back((*i).block_index);
+            }
+        }
+
+        for (state_type i = 0; i < aut.num_states(); i++) 
+        {
+            block_index_type bid = block_index_of_a_state[i];
+            if (!block_flags[bid]) {
+                block_flags[bid] = true; 
+                partition.insert(bid);
+            }
+        }
+        mCRL2log(mcrl2::log::info) << "Partition refinement done, partition contains: " << partition.size() << " blocks, the history contains "<< lvl-1 << " levels." <<std::endl;
+        save_transitions();
+    }
+
+    /** \brief Destroys this partitioner. */
+    ~bisim_partitioner_counter_example() = default;
+
+    /** \brief Creates a state formula that distinguishes state s from state t.
+     *  \details The states s and t are non bisimilar states. A distinguishign state formula phi is
+     *           returned, which has the property that s \in \sem{phi} and  t \not\in\sem{phi}.
+     *           Based on the preprint "Computing minimal distinguishing Hennessey-Milner formulas is NP-hard
+     *           . But variants are tractable", 2023 by Jan Martens and Jan Friso Groote
+     *  \param[in] s The state number for which the resulting formula should be true
+     *  \param[in] t The state number for which the resulting formula should be false
+     *  \return A minimal observation depth distinguishing state formula, that is often also minimum negation-depth and irreducible. */
+    mcrl2::state_formulas::state_formula dist_formula_mindepth(const std::size_t s, const std::size_t t) {
+        formula f = distinguish(block_index_of_a_state[s], block_index_of_a_state[t]);
+        mCRL2log(mcrl2::log::info) << "done with formula \n";
+        return convertFormula(f);
```

Comment:
I don't know what the state is of the other bisimulation algorithms and I should publish this auto formatting file as some point, but I think here snake_case is mostly preferred for function names (convert_formula).